src/utils/work/fetchWork.js

```diff
@@ -1,59 +1,51 @@
 import { Environment, isBrowser } from 'utils/deploymentInfo';
 
-import { getBackendStatus } from 'redux/selectors';
-
 import cache from 'utils/cache';
-import generateETag from 'utils/work/generateETag';
-import { dispatchWorkRequest, seekFromS3 } from 'utils/work/seekWorkResponse';
-
-// Temporarily using gene expression without local cache
-const fetchGeneExpressionWorkWithoutLocalCache = async (
-  experimentId,
+import dispatchWorkRequest from 'utils/work/dispatchWorkRequest';
+import downloadFromS3 from 'utils/work/downloadFromS3';
+import waitForWorkRequest from './waitForWorkRequest';
+
+// retrieveData will try to get the data for the given experimentId and ETag from
+// the fastest source possible. It will try to get the data in order from:
+// 1. Browser cache
+// 2. S3 (cache) via signedURL
+// 3. Worker
+//   3.1. Via socket (small data)
+//   3.2. Via S3 (large data) via signedURL
+const retrieveData = async (experimentId,
+  ETag,
+  signedUrl,
+  request,
   timeout,
   body,
-  backendStatus,
-  environment,
-  broadcast,
-  extras,
-  onETagGenerated,
   dispatch,
-  getState,
-) => {
-  // If new genes are needed, construct payload, try S3 for results,
-  // and send out to worker if there's a miss.
-  const { pipeline: { startDate: qcPipelineStartDate } } = backendStatus;
-
-  const ETag = await generateETag(
-    experimentId,
-    body,
-    extras,
-    qcPipelineStartDate,
-    environment,
-    dispatch,
-    getState,
-  );
+  useCache) => {
+  // 1. Check if we have the ETag in the browser cache (no worker)
+  const cachedData = useCache ? await cache.get(ETag) : null;
+  if (cachedData) {
+    return cachedData;
+  }
 
-  onETagGenerated(ETag);
+  // 2. Check if data is cached in S3 so we can download from the signed URL (no worker)
+  if (signedUrl) {
+    return await downloadFromS3(body.name, signedUrl);
+  }
 
-  try {
-    const { signedUrl, data } = await dispatchWorkRequest(
-      experimentId,
-      body,
-      timeout,
-      ETag,
-      {
-        ETagPipelineRun: qcPipelineStartDate,
-        broadcast,
-        ...extras,
-      },
-      dispatch,
-    );
+  // 3. If we don't have signedURL, wait for the worker to send us the data via
+  // - the data via socket
+  // - the signedURL to download the data from S3
+  const { signedUrl: workerSignedUrl, data } = await waitForWorkRequest(ETag,
```

Comment:
What we usually do is having all the params at hte same level like this:

```
await waitForWorkRequest(
  ETag,
  experimentId,
  request,
  timeout,
  dispatch
);
```

Reply:
I feel like the linter should have noticed this, it auto corrects it for me. Do you have it set up?